dbms/src/Functions/repeat.cpp

```diff
@@ -0,0 +1,236 @@
+#include <Functions/castTypeToEither.h>
+
+namespace DB
+{
+	namespace ErrorCodes
+	{
+		extern const int ILLEGAL_COLUMN;
+		extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+	}
+
+	struct RepeatImpl
+	{
+        static void vectorNonConstStr(
+            const ColumnString::Chars & data,
+            const ColumnString::Offsets & offsets,
+            ColumnString::Chars & res_data,
+            ColumnString::Offsets & res_offsets,
+            const UInt64 repeatTime)
+        {
+            UInt64 data_size = 0;
+            res_offsets.assign(offsets);
+            for (UInt64 i = 0; i < offsets.size(); ++i)
+			{
+                data_size += (offsets[i] - offsets[i - 1] - 1) * repeatTime + 1;
+                res_offsets[i] = data_size;
+            }
+            res_data.resize(data_size);
+            for (UInt64 i = 0; i < res_offsets.size(); ++i)
+            {
+                array(data.data() + offsets[i - 1], res_data.data() + res_offsets[i - 1], offsets[i] - offsets[i - 1], repeatTime);
+            }
+        }
+
+        static void
+        vectorConst(const String & copy_str, const UInt64 repeatTime, ColumnString::Chars & res_data, ColumnString::Offsets & res_offsets)
+        {
+            UInt64 data_size = copy_str.size() * repeatTime + 1;
+            res_data.resize(data_size);
+            res_offsets.resize_fill(1, data_size);
+            array((UInt8 *)copy_str.data(), res_data.data(), copy_str.size() + 1, repeatTime);
+        }
+
+        template <typename T>
+        static void vectorNonConst(
+            const ColumnString::Chars & data,
+            const ColumnString::Offsets & offsets,
+            ColumnString::Chars & res_data,
+            ColumnString::Offsets & res_offsets,
+            const PaddedPODArray<T> & col_num)
+        {
+            UInt64 data_size = 0;
+            res_offsets.assign(offsets);
+            for (UInt64 i = 0; i < col_num.size(); ++i)
+			{
+                data_size += (offsets[i] - offsets[i - 1] - 1) * col_num[i] + 1;
+                res_offsets[i] = data_size;
+            }
+            res_data.resize(data_size);
+            for (UInt64 i = 0; i < col_num.size(); ++i)
+			{
+                array(data.data() + offsets[i - 1], res_data.data() + res_offsets[i - 1], offsets[i] - offsets[i - 1], col_num[i]);
+            }
+        }
+
+
+        template <typename T>
+        static void vectorNonConstInteger(
+            const String & copy_str, ColumnString::Chars & res_data, ColumnString::Offsets & res_offsets, const PaddedPODArray<T> & col_num)
+        {
+            UInt64 data_size = 0;
+            res_offsets.resize(col_num.size());
+            UInt64 str_size = copy_str.size();
+            for (UInt64 i = 0; i < col_num.size(); ++i)
+			{
+                data_size += str_size * col_num[i] + 1;
+                res_offsets[i] = data_size;
+            }
+            res_data.resize(data_size);
+            for (UInt64 i = 0; i < col_num.size(); ++i)
```

Comment:
Calculation of col_num.size() should be done outside of the loop (outside it may fail to optimize due to aliasing issues).